### ISSUE TITLE
Two data sets share a label input network training How to change the network

### ISSUE THREAD

**Issue by wq247726404** [ultralytics/ultralytics#8213]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

![image](https://github.com/ultralytics/ultralytics/assets/40755124/d6142e84-b38e-48bf-bbd3-3e00e5c198f4)

**Comment by wq247726404** [comment]
> @wq247726404 hello! If you're looking to train a network on two datasets that share a common label, you can simply merge the datasets and train your model on the combined set. Here's a quick example of how you might organize your data and modify your dataset YAML file:
> 
> ```yaml
> # datasets.yaml
> train: [dataset1/train, dataset2/train]  # Paths to train images
> val: [dataset1/val, dataset2/val]        # Paths to validation images
> 
> # Your common label names file
> names: ['shared_label', ...]  # List of class names
> ```
> 
> Make sure your annotations for both datasets are consistent with the names file. Then, you can train your model as usual:
> 
> ```shell
> yolo train data=datasets.yaml model=yolov8n.yaml
> ```
> 
> Happy training! 😊🚀

Thank you for your guidance, but according to the method you said, only dataset2 participated in the training

**Comment by glenn-jocher** [comment]
@wq247726404 if you have two datasets that share a label and you want to train a network on both datasets, you can merge the datasets and create a combined dataset with a unified set of labels. Here's how you can approach this:

1. **Merge the Datasets**: Combine the images and annotations from both datasets into a single dataset. Make sure that the label that is shared between the two datasets has the same class index in the combined dataset.

2. **Create a Unified Data YAML**: Create a new `data.yaml` file that points to the paths of the combined training and validation images and annotations. The `data.yaml` file should also list all the unique class labels used in the combined dataset.

3. **Modify the Model Configuration**: If necessary, adjust the model configuration file (e.g., `yolov8n.yaml`) to match the number of classes in the combined dataset. The `nc` (number of classes) parameter in the model configuration should reflect the total number of unique classes.

4. **Train the Network**: Use the combined dataset and the updated `data.yaml` and model configuration files to train the network.

Here's an example of what the `data.yaml` might look like for the combined dataset:

```yaml
train: /path/to/combined/dataset/train/images
val: /path/to/combined/dataset/val/images

nc: 80  # Number of classes
names:
  - class1
  - class2
  - class3
  # ... add all class names
```

And an example of the model configuration adjustment in `yolov8n.yaml`:

```yaml
# parameters
nc: 80  # number of classes

# model architecture details
# ...
```

Once you have the combined dataset and the updated configuration files, you can start training with the following command:

```bash
yolo detect train data=combined_data.yaml model=yolov8n.yaml
```

Remember to replace `combined_data.yaml` with the path to your actual combined `data.yaml` file and `yolov8n.yaml` with the path to your actual model configuration file.

If you encounter any issues or have specific requirements, please provide more details so I can offer more targeted assistance.

**Comment by wq247726404** [comment]
I'm sorry I wasn't clear. I want to use two different types of data sets, share a label, the two data sets and labels are one-to-one correspondence, can not be mixed in a folder, is changed into two inputs, and then input the network at the same time to extract features after feature fusion.



---Original---
From: "Glenn ***@***.***&gt;
Date: Fri, Feb 16, 2024 00:56 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [ultralytics/ultralytics] Two data sets share a label inputnetwork training How to change the network (Issue #8213)




 
@wq247726404 if you have two datasets that share a label and you want to train a network on both datasets, you can merge the datasets and create a combined dataset with a unified set of labels. Here's how you can approach this:
  
 
Merge the Datasets: Combine the images and annotations from both datasets into a single dataset. Make sure that the label that is shared between the two datasets has the same class index in the combined dataset.
 
 
 
Create a Unified Data YAML: Create a new data.yaml file that points to the paths of the combined training and validation images and annotations. The data.yaml file should also list all the unique class labels used in the combined dataset.
 
 
 
Modify the Model Configuration: If necessary, adjust the model configuration file (e.g., yolov8n.yaml) to match the number of classes in the combined dataset. The nc (number of classes) parameter in the model configuration should reflect the total number of unique classes.
 
 
 
Train the Network: Use the combined dataset and the updated data.yaml and model configuration files to train the network.
 
  
Here's an example of what the data.yaml might look like for the combined dataset:
 train: /path/to/combined/dataset/train/images val: /path/to/combined/dataset/val/images nc: 80  # Number of classes names:   - class1   - class2   - class3   # ... add all class names
 
And an example of the model configuration adjustment in yolov8n.yaml:
 # parameters nc: 80  # number of classes # model architecture details # ...
 
Once you have the combined dataset and the updated configuration files, you can start training with the following command:
 yolo detect train data=combined_data.yaml model=yolov8n.yaml
 
Remember to replace combined_data.yaml with the path to your actual combined data.yaml file and yolov8n.yaml with the path to your actual model configuration file.
 
If you encounter any issues or have specific requirements, please provide more details so I can offer more targeted assistance.

**Comment by glenn-jocher** [comment]
@wq247726404 when training a network on two datasets that share a common label, you can simply merge the datasets and train the network on the combined set. However, if you need to distinguish between the same label from different datasets, you might consider one of the following approaches:

1. **Rename Labels**: If you want the network to differentiate between the same label from different datasets, you could rename the labels to reflect their origin. For example, if both datasets have the label "car", you could rename them to "dataset1_car" and "dataset2_car".

2. **Modify Dataset Configuration**: In your dataset YAML file, you can specify different paths for the training and validation images and labels. If you're merging datasets, ensure that the `nc` (number of classes) parameter reflects the total unique labels across both datasets.

Here's an example of how you might structure your dataset YAML file after renaming labels to distinguish them:

```yaml
# dataset1_and_dataset2.yaml
path: ../datasets/merged-dataset  # common base path for both datasets
train: [dataset1/train/images, dataset2/train/images]  # training images from both datasets
val: [dataset1/val/images, dataset2/val/images]  # validation images from both datasets
test:  # optional test images

# Classes
nc: 20  # number of classes (total unique labels from both datasets)
names: ['dataset1_label1', 'dataset1_label2', ..., 'dataset2_label1', 'dataset2_label2', ...]  # class names
```

3. **Custom Data Loader**: If you need more control over how data from different datasets is loaded and processed, you might need to write a custom data loader that handles the specifics of your datasets.

4. **Multi-Task Learning**: If your goal is to perform multi-task learning where the network learns to perform different tasks for different datasets, you would need to modify the network architecture to have multiple heads, one for each task, and adjust the loss functions accordingly.

5. **Transfer Learning**: If the datasets are significantly different and you want to leverage knowledge from one dataset to another, you could perform transfer learning. Train the network on one dataset, then fine-tune it on the other dataset.

Remember to adjust your training script to use the new dataset configuration file. If you're using the Ultralytics YOLOv8 repository, you can specify the dataset configuration file when you initiate training using the `data` argument:

```bash
yolo train data=dataset1_and_dataset2.yaml model=yolov8n.yaml
```

Lastly, the image link you provided seems to be broken. If you need to share an image, please ensure the link is correct and publicly accessible.
